crates/holochain_sqlite/tests/db_wrapper.rs

```diff
@@ -0,0 +1,383 @@
+    for _ in 0..num_readers {
+        let c = db_handle.conn_permit::<DatabaseError>().await.unwrap();
+        held_permits.push(c)
+    }
+
+    assert_eq!(num_readers, held_permits.len());
+}
+
+// TODO this test shows that waiting for a write permit will block until another one is released, whether or not
+//      that is possible.
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[tokio::test]
+async fn single_write_permit() {
+    holochain_trace::test_run().unwrap();
+
+    let tmp_dir = tempfile::TempDir::new().unwrap();
+    let db_handle = DbWrite::open(&tmp_dir.into_path(), TestDatabaseKind::new()).unwrap();
+
+    let _hold_permit = db_handle.conn_write_permit().await;
+
+    // Bad, will never stop waiting!
+    let result = tokio::time::timeout(
+        std::time::Duration::from_secs(35),
+        db_handle.conn_write_permit(),
+    )
+    .await;
+
+    assert!(result.is_err());
+    // The inner cannot error so this is a timeout error
+    assert!(result.err().is_some())
+}
+
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[tokio::test]
+async fn write_permits_can_be_released() {
+    holochain_trace::test_run().unwrap();
+
+    let tmp_dir = tempfile::TempDir::new().unwrap();
+    let db_handle = DbWrite::open(&tmp_dir.into_path(), TestDatabaseKind::new()).unwrap();
+
+    let hold_permit = db_handle.conn_write_permit().await;
+
+    drop(hold_permit);
+
+    // Does not return a result so we can't do much to check the given permit is valid. If the test exits instead of deadlocking then it passed...
+    db_handle.conn_write_permit().await;
+}
+
+// TODO The database wrapper is a leaky abstraction around the database pool. Being able to get a permit does NOT mean
+//      you can actually access the database.
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[tokio::test]
+async fn can_get_a_write_permit_when_the_pool_is_exhausted() {
+    holochain_trace::test_run().unwrap();
+
+    let tmp_dir = tempfile::TempDir::new().unwrap();
+    let db_handle = DbWrite::open(&tmp_dir.into_path(), TestDatabaseKind::new()).unwrap();
+
+    let num_readers = num_read_threads() + 1;
+
+    let mut held_connections = Vec::with_capacity(num_readers);
+    for _ in 0..num_readers {
+        let c = db_handle.conn().unwrap();
+        held_connections.push(c)
+    }
+
+    let writer_permit = db_handle.conn_write_permit().await;
+
+    // Now we have a write permit but all connections are held? So although we're the only 'writer' we can't write
+    assert_eq!(num_readers, held_connections.len());
+
+    let result = db_handle.with_permit(writer_permit);
+
+    // and get an error trying to retrieve a connection using the permit
+    assert!(result.is_err());
+}
+
+// TODO This is bad but really a consequence of being able to claim and hold write permits. Also because the attempt to
+//      get a write permit internally will never give up. The only remaining problem is if the
```

Comment:
> The only remaining problem is if the

what's the only remaining problem... i must know!

Reply:
What a cliffhanger!

Reply:
Sorry about this, I jumped around a bit while writing these comments as my understanding of the code developed. Luckily I can still remember what I was writing. I've pushed the update but for reference here it should have said

> The only remaining problem is if the query itself does something it shouldn't like waiting on a network call or anything else slow because we don't prevent the query itself from blocking for unreasonably long.

If the passed function does not finish in a reasonable amount of time because it does something other than running a single database query then it holds onto the only available writer permit for that time. I haven't checked whether this is the case anywhere but I do want to keep that thought around as something to be able to at least measure so we know if it becomes a problem.